continuous_queries.md

```diff
@@ -0,0 +1,236 @@
+
+Unfortunately, there's no easy way to tell cluster wide if there were data points written in an already passed window of time for a CQ. We might be able to add this at the data nodes and have them track it, but it would be quite a bit more work.
+
+The easy way would just be to have CQs re-execute for periods that recently passed and have some user-configurable window of time that they stop checking after. Then we could give the user the ability to recalculate CQs ranges of time if they need to correct for some problem that occurred or the loading of a bunch of historical data.
+
+With this approach, we'd have the metadata in the database store the last time each CQ was run. Whenever the Raft leader sent out a command to a data node to handle a CQ, the data node would use this metadata to determine which windows of time it should compute.
+
+This approach is like what exists in 0.8, with the exception that it will automatically catch data that is lagged behind in a small window of time and give the user the ability to force recalculation.
+
+#### Expanding writes
+
+When a write comes into a data node, we could have it evaluated against group by CQs in addition to the non-group by ones. It would then create writes that would then go through the brokers. When the CQ writes arrive at the data nodes, they would have to handle each write differently depending on if it was a write to a raw series or if it was a CQ write.
+
+Let's lay out a concrete example.
+
+```sql
+CREATE CONTINUOUS QUERY 10m_cpu_by_region
+ON foo
+BEGIN
+  SELECT mean(value)
+  INTO cpu_by_region
+  FROM cpu
+  GROUP BY time(10m), region
+END
+```
+
+In this example we write values into `cpu` with the tags `region` and `host`.
+
+Here's another example CQ:
+
+```sql
+CREATE CONTINUOUS QUERY 1h_cpu
+ON foo
+BEGIN
+  SELECT mean(value)
+  INTO 1h.cpu
+  FROM raw.cpu
+  GROUP BY time(10m), *
+END
+```
+
+That would output one series into the `1h` retention policy for the `cpu` measurement for every series from the `raw` retention policy and the `cpu` measurement.
+
+Both of these examples would be handled the same way despite one being a big merge of a bunch of series into one and the other being an aggregation of series in a 1-to-1 mapping.
+
+Say we're collecting data for two hosts in a single region. Then we'd have two distinct series like this:
+
+```
+1 - cpu host=serverA region=uswest
+2 - cpu host=serverB region=uswest
+```
+
+Whenever a write came into a server, we'd look at the continuous queries and see if we needed to create new writes. If we had the two CQ examples above, we'd have to expand a single write into two more writes (one for each CQ).
+
+The first CQ would have to create a new series:
+
+```
+3 - cpu_by_region region=uswest
+```
+
+The second CQ would use the same series id as the write, but would send it to another retention policy (and thus shard).
```

Comment:
That's right. Because the measurement name and tagset are the same, which are what a series ID maps to.\